File: F#/F#/queue.fs
module FSharp.Queue


/// bad implementation of a queue with O(n) dequeue
type Queue<'T>() =
    let data = ResizeArray<'T>()
    member this.size = data.Count
    member this.enqueue item = data.Add item
    member this.dequeue() =
        let res = data.[0]
        data.RemoveAt 0
        res
    member this.isEmpty = data.Count = 0

let hotPotato names num =
    let queue = Queue()
    List.iter queue.enqueue names

    while queue.size > 1 do
        for i = 1 to num do
            queue.enqueue(queue.dequeue())
        queue.dequeue() |> ignore

    queue.dequeue()

File: Java/src/Queue/Queue.java
package Queue;

import java.util.ArrayList;

/**
 * Created by Daniel on 4/5/2015.
 * Bad implementation of a queue with O(n) dequeue.
 */
public class Queue<T> {
    private final ArrayList<T> data;

    public Queue() {
        this.data = new ArrayList<>();
    }

    public boolean isEmpty(){
        return data.isEmpty();
    }

    public void enqueue(T i) {
        data.add(i);
    }

    public int size() {
        return data.size();
    }

    public T dequeue() {
        T res = data.get(0);
        data.remove(0);
        return res;
    }
}
